### ISSUE TITLE
nfsmount broken in 1.68.0 beta

### ISSUE THREAD

**Issue by claudio-tinnirello-filescom** [rclone/rclone#8029]
#### What is the problem you are having with rclone?

The nfsmount command doesn't work properly in version 1.68.0 of rclone. It mounts the remote but, in the Finder, shows certain subfolders and as having "no entry" icons. Those subfolders cannot be accessed. Also, some files are unable to be accessed.

The mount itself seems to work, but items within the mount are unaccessible, even if I have full permissions on the remote.

Trying to list these subfolders from the command line gives the following error:

```
ls: ./ERP: Stale NFS file handle
```

When trying to access these items from Mac Finder, a popup occurs with the following message: `The folder “ERP” can’t be opened because you don’t have permission to see its contents.`

When using version 1.67.0, on the exact same remote, with the exact same authentication options, there are none of these issues.

#### What is your rclone version (output from `rclone version`)

1.68.0

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

MacOS Sonoma 14.6.1

#### Which cloud storage system are you using? (e.g. Google Drive)

Reproduced with AWS S3 and Pcloud type remotes.

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone --vfs-cache-mode writes --cache-dir ./cache nfsmount aws: ./mount1  --allow-other --default-permissions

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752891 (nfs.GetAttr)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=D90872F95AB048679EB4CF0444C8F98B
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=<nil>
2024/08/21 10:00:54 DEBUG : : Lstat:
2024/08/21 10:00:54 DEBUG : : >Lstat: fi=/, err=<nil>
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752892 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752893 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752894 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752895 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752896 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752897 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752898 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752899 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752900 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752901 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:54 DEBUG : [NFS DEBUG] request: RPC #1628752902 (nfs.Access)
2024/08/21 10:00:54 DEBUG : nfs: FromHandle: handle=0F83C09216C342ED92C1516F387EB179
2024/08/21 10:00:54 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/21 10:00:56 DEBUG : [NFS DEBUG] request: RPC #1628752903 (nfs.GetAttr)
2024/08/21 10:00:56 DEBUG : nfs: FromHandle: handle=36AC9F2B18094C61AFFA050948BCE103
2024/08/21 10:00:56 DEBUG : nfs: >FromHandle: path=["filescom-product" "Claudio"], err=<nil>
2024/08/21 10:00:56 DEBUG : filescom-product/Claudio: Lstat:
2024/08/21 10:00:56 DEBUG : filescom-product/Claudio: >Lstat: fi=filescom-product/Claudio/, err=<nil>
2024/08/21 10:01:01 DEBUG : [NFS DEBUG] request: RPC #1628752904 (nfs.GetAttr)

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by kapitainsky** [comment]
joy of beta testing:)

There is ongoing work on `nfsmount` with some changes not committed to main branch yet.

Can you try https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids/

which sorts out some ownership/premission problems.

**Comment by ncw** [comment]
I did notice this while investigation #7973 but I thought it was just Linux NFS being a bit flaky.

The OS caches the NFS handles (which are like an identifier for files) and between the beta and v1.67.0 we changed the way those handles are made.

I suspect you were running v1.67.0 then upgraded to the beta which means the OS had the old handles cached.

So you might find a reboot fixes the problem - it would be helpful if you could try that.

I am trying to make the nfs server robust in general for v1.68 not just for macOS so this is a very timely report.

**Comment by claudio-tinnirello-filescom** [comment]
I tried https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids/ and had the exact same problems.  The debug log entries also look the same:

```
2024/08/22 10:37:32 DEBUG : : Lstat:
2024/08/22 10:37:32 DEBUG : : >Lstat: fi=/, err=<nil>
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141070 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141071 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141072 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141073 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141074 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141075 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141076 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141077 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141078 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141079 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:34 DEBUG : [NFS DEBUG] request: RPC #2004141080 (nfs.Access)
2024/08/22 10:37:34 DEBUG : nfs: FromHandle: handle=51347680B3E549A0B3C2FE9165921D04
2024/08/22 10:37:34 DEBUG : nfs: >FromHandle: path=[], err=Invalid file handle
2024/08/22 10:37:37 DEBUG : [NFS DEBUG] request: RPC #2004141081 (nfs.GetAttr)
2024/08/22 10:37:37 DEBUG : nfs: FromHandle: handle=2A27BA72EDAC4B6CAB98FC3A5BF1F4EB
2024/08/22 10:37:37 DEBUG : nfs: >FromHandle: path=[], err=<nil>
```

Yes, I've been comparing 1.67 with 1.68 but rebooting made no difference between them.

Also, the `mount` command is missing in the 1.68.0 beta that I was using.

Trying it in the build you pointed me at (above) gives an error:

`2024/08/22 10:37:57 Fatal error: failed to mount FUSE fs: mount stopped before calling Init: mount failed: cgofuse: cannot find FUSE`

I'll re-read the docs as I'm probably missing FUSE on my machine.

**Comment by claudio-tinnirello-filescom** [comment]
Both macFUSE and FUSE-T require administrator privileges to install, which I don't have on my Mac. Looks like the `mount` command isn't an option for me.

**Comment by claudio-tinnirello-filescom** [comment]
Thanks for the pointer to the release notes! The main page didn't give this info and trying to install from the PKG was also asking for admin permissions.

I managed to get fuse-t installed, but it doesn't work right away.

For others trying this, I also had to create a symlink:

`ln -s $HOME/Library/Application\ Support/fuse-t/lib/libfuse-t-1.0.38.dylib /usr/local/lib/libfuse-t.dylib`

Once the symlink was in place, I was able to run the `rclone mount` command and confirm that the permission issues I reported went away.

I'm beta testing the Files.com integration but was unable to test that with this branch. I'm assuming that if the permission problems went away for aws and pcloud then they'll also go away for Files.com.

Thanks for your help!

**Comment by kapitainsky** [comment]
> For others trying this, I also had to create a symlink:

Brilliant! Thank you for sharing details.

**Comment by ncw** [comment]
@kapitainsky can you reproduce the original issue here? Maybe swap between v1.67.0 and [the latest beta](https://beta.rclone.org/)? I can't reproduce it on linux at all any more. I'm worried I've done something wrong with the NFS handle changes I did but I can't figure out what.

Thank you :-)

**Comment by kapitainsky** [comment]
> @kapitainsky can you reproduce the original issue here?

Yes something does not work with the latest beta:

```
$ rclone-beta nfsmount OD01-ENC: ~/Temp/mount --vfs-cache-mode=full -vv
2024/08/27 09:59:40 DEBUG : rclone: Version "v1.68.0-beta.8241.0e5eb7a9b" starting with parameters ["rclone-beta" "nfsmount" "OD01-ENC:" "/Users/kptsky/Temp/mount" "--vfs-cache-mode=full" "-vv"]
2024/08/27 09:59:40 DEBUG : Creating backend with remote "OD01-ENC:"
...
```

Some (random? I can not see any pattern) directories are not accessible:

```
$ cd ~/Temp/mount/arc_rcc
~/Temp/mount/arc_rcc$ cd RPi-images/
-bash: cd: RPi-images/: Stale NFS file handle
```

It all works with rclone v1.67 (but using MS Office #7973)

**Comment by ncw** [comment]
Thanks for the confirmation @kapitainsky and the new info that it could be different directories. I've obviously broken something subtle here so will try harder to replicate and fix.

**Comment by ncw** [comment]
@kapitainsky I can't replicate this :-(

Can you do a bisect for me to try to find the commit which causes the problem? It should be using the same code to generate the file handles so I don't understand why it is breaking. It might be that ce7dfa075c9286bf73d579f81d6f40f3fd1a01ec which upgraded the library is actually the problem, or it might be 70e8ad456fcc627f57c001aa82ac925bd9fb47a8 or 55b9b3e33a1c98740a6e5994ec093b2b553d7458 or something else entirely!

**Comment by kapitainsky** [comment]
> Can you do a bisect for me to try to find the commit which causes the problem?

Here you are:

```
28ba4b832d009133d51d7dc742040f07b590c8b0 is the first bad commit
commit 28ba4b832d009133d51d7dc742040f07b590c8b0 (HEAD)
Author: Nick Craig-Wood <nick@craig-wood.com>
Date:   Wed Jul 3 19:20:43 2024 +0100

    serve nfs: convert options to new style

 cmd/serve/nfs/nfs.go | 24 ++++++++++++++++++------
 1 file changed, 18 insertions(+), 6 deletions(-)
```

**Comment by ncw** [comment]
Thank you @kapitainsky - exactly what I needed.

I've been testing this with `rclone serve nfs` but the problem is with `rclone nfsmount` as became clear after investigating that commit.

Can you and @claudio-tinnirello-filescom give this a try please?

[v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

**Comment by kapitainsky** [comment]
> Can you and @claudio-tinnirello-filescom give this a try please?
> 
> [v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

```
$ git rev-parse --short HEAD
7771456cb
```

works now. All dirs are accessible again as with rclone v1.67

**Comment by claudio-tinnirello-filescom** [comment]
Tested again with [v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8242.7771456cb.fix-7973-nfs-server-ids/) and the permission problems have gone away.

Tested against AWS S3, Pcloud, and Files.com and all worked as expected.

Pcloud was giving me issues when trying to open files but the others were fine so I assume that this is a Pcloud specific issue and I'm not worried about it.

Great work guys!

**Comment by ncw** [comment]
Thank you both for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.68
